src/NuGet.Core/NuGet.Versioning/IVersionRangeComparer.cs

```diff
@@ -8,7 +8,7 @@ namespace NuGet.Versioning
     /// <summary>
     /// Version range comparer
     /// </summary>
-    public interface IVersionRangeComparer : IEqualityComparer<VersionRangeBase>
+    public interface IVersionRangeComparer : IEqualityComparer<VersionRangeBase?>
```

Comment:
@zivkan this change (and another like it) were the reason I had to make `IEqualityComparer.GetHashCode`'s argument nullable.

There was another change that required this type parameter to be nullable, but I can't find it in a browser. If you check out this branch and remove this one change, you'll find it.

Reply:
I checked out your branch on my devbox, changed `IEqualityComparer<VersionRangeBase?>` back to `IEqualityComparer<VersionRangeBase>`, and it builds without error or warning on my machine.

Reply:
Maybe it was the other instance of this that required it. I'll take a look on my PC in a bit, as I'm about to start my work day.

Reply:
Yeah it was the other instance, and I copied that here for consistency.

Changing `IVersionComparer` to remove both `?` causes compile errors:

> src\NuGet.Core\NuGet.Versioning\SemanticVersionBase.cs(134,43,134,48): error CS8604: Possible null reference argument for parameter 'y' in 'int IComparer<SemanticVersion>.Compare(SemanticVersion x, SemanticVersion y)'.
>src\NuGet.Core\NuGet.Versioning\SemanticVersionBase.cs(125,42,125,47): error CS8604: Possible null reference argument for parameter 'y' in 'bool IEqualityComparer<SemanticVersion>.Equals(SemanticVersion x, SemanticVersion y)'.

They both handle nulls correctly, so changing the advertised type seems fair to me.

Reply:
I understand what you're saying. It still didn't compile without that change however.

Reply:
In my PR when I added null annotations to the assembly, I ignored a bunch of warnings, because the .NET Framework is missing annotations and the compiler incorrectly assumes that nulls are not allowed: https://github.com/NuGet/NuGet.Client/pull/5097/files

https://github.com/NuGet/NuGet.Client/blob/4b6a87bfe90d247f961eba41d63dbf418a56fdcc/src/NuGet.Core/NuGet.Versioning/FloatRange.cs#L442-L445

Unsurprisingly, since I added these suppressions in my own PR, I think it would be better for you to add the same suppressions in your PR. This way in the future when we can drop .NET Framework, we're not left with API design bugs.

Reply:
I'm happy either way. I'm content with annotating the methods to reflect the values they actually accept, though I don't know why `GetHashCode` accepts a null. It seems like the alternative means every call site needs to know it's fine to suppress the warning, but if that's preferred then that works. Does 3b08080a5e26aac25531bc68d04aedbddd7d12a7 capture what you had in mind?